### PR TITLE
Introduce AccessLogWriterUtil and Remove Unnecessary RequestContext Push/Pop

### PR DESCRIPTION
**Pull request by yzfeng2020**
Motivation:

As described in https://github.com/line/armeria/issues/5984, there are instances of unnecessary request context push/pop operations related to access logging when the access log writer is disabled. These operations are redundant since the logging path does not need to be triggered when the writer is disabled. Optimizing this behavior reduces unnecessary overhead in such cases.

Modifications:

- Added `AccessLogWriterUtil#maybeWriteAccessLog`
    - Writes access log only when `TransientServiceOption#WITH_ACCESS_LOGGING` is enabled and the access log writer is not disabled.
- Updated `AbstractHttpResponseHandler` and `HttpServerHandler`

Result:

- Closes https://github.com/line/armeria/issues/5984.
- Removes unnecessary context push/pop operations for access logging when the access log writer is disabled, resulting in a more efficient request handling process.

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java [hunk 1]
```diff
@@ -50,7 +55,15 @@ static AccessLogWriter combined() {
...
     static AccessLogWriter disabled() {
-        return requestLog -> { /* No operation. */ };
+        return DISABLED;
+    }
+
+    /**
+     * Returns whether this writer is the disabled access log writer.
+     *
+     */
+    default boolean isDisabled() {
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java:65]: How about doing `accessLogWriter != AccessLogWriter.DISABLED` this instead of introducing this new public API?

**Review by minwoox — APPROVED** [review]
Looks good. Thanks!

File: core/src/main/java/com/linecorp/armeria/server/ServiceConfig.java [hunk 2]
```diff
@@ -474,6 +474,16 @@ List<ShutdownSupport> shutdownSupports() {
...
 
+    /**
+     * Returns whether the access log should be written for this {@link #service()}.
+     * Access log is written only when the {@link TransientServiceOption#WITH_ACCESS_LOGGING} option is enabled
+     * and the {@link #accessLogWriter()} is not {@link AccessLogWriter#DISABLED}.
+     */
+    public boolean shouldWriteAccessLog() {
+        return !accessLogWriter.isDisabled() &&
+                transientServiceOptions.contains(TransientServiceOption.WITH_ACCESS_LOGGING);
+    }
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/ServiceConfig.java:485]: Should we create `AccessLogWriterUtil` and move this method to it? I don't think this ㅡethod needs to be a public API.

File: core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java (hunk 1 above)
Comment (yzfeng2020) [reply core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java]: done!

File: core/src/main/java/com/linecorp/armeria/server/ServiceConfig.java (hunk 2 above)
Comment (yzfeng2020) [reply core/src/main/java/com/linecorp/armeria/server/ServiceConfig.java]: sg, I moved the `maybeWriteAccessLog` entirely

**Review by jrhee17 — APPROVED** [review]
👍 👍

File: core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java [hunk 3]
```diff
@@ -50,7 +56,7 @@ static AccessLogWriter combined() {
      * Returns disabled access log writer.
      */
     static AccessLogWriter disabled() {
-        return requestLog -> { /* No operation. */ };
```
Comment (ikhoon) [inline core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java:53]: I belive we can revert this change because Java returns a singleton instance for a lambda that does not capture a variable outside. 
In addition, conventionally, we don't add a field in an `interface` to be free from breaking changes.

Comment (yzfeng2020) [reply core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java:53]: A test shows me otherwise

```java
    @Test
    void testAccessLogger() {
        final AccessLogWriter accessLogWriter1 = log -> {
        };

        final AccessLogWriter accessLogWriter2 = log -> {
        };

        assert(accessLogWriter1 == accessLogWriter2);   <--- false
    }
```

@ikhoon what is your recommendation? creating a public DisbaledAccessLogWriter class?

Comment (ikhoon) [reply core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java:53]: `accessLogWriter1` and `accessLogWriter2` are different objects.
I meant `AccessLogWriter.disabled()` always returns the same instance.
```java
assert AccessLogWriter.disabled() == AccessLogWriter.disabled();
```

Comment (yzfeng2020) [reply core/src/main/java/com/linecorp/armeria/server/logging/AccessLogWriter.java]: thanks for the clarification, updated!
